test/common/src/main/kotlin/org/codice/compliance/verification/core/CoreVerifier.kt

```diff
@@ -15,17 +15,47 @@ package org.codice.compliance.verification.core
 
 import org.apache.cxf.rs.security.saml.sso.SSOConstants.SIGNATURE
 import org.codice.compliance.SAMLComplianceException
+import org.codice.compliance.SAMLSpecRefMessage
 import org.codice.compliance.SAMLSpecRefMessage.SAMLCore_5_4_1
 import org.codice.compliance.SAMLSpecRefMessage.SAMLCore_5_4_2_a
 import org.codice.compliance.SAMLSpecRefMessage.SAMLCore_5_4_2_b
 import org.codice.compliance.SAMLSpecRefMessage.SAMLCore_5_4_2_b1
 import org.codice.compliance.SAMLSpecRefMessage.SAMLCore_6_1_b
 import org.codice.compliance.allChildren
 import org.codice.compliance.children
+import org.codice.compliance.utils.TestCommon
 import org.codice.compliance.utils.TestCommon.Companion.ELEMENT
 import org.w3c.dom.Node
 
 class CoreVerifier(val node: Node) {
+
+    /**
+     * Verifies that the response has an appropriate error status code.
+     * This should be called explicitly if an error is expected
+     */
+    fun verifyErrorStatusCode(samlErrorCode: SAMLSpecRefMessage) {
+        val statusCodes = node.allChildren("StatusCode")
+        if (statusCodes.isEmpty())
+            throw SAMLComplianceException.createWithPropertyMessage(samlErrorCode,
+                    property = "Status Codes",
+                    actual = "[]",
+                    expected = "<StatusCode Value=...>")
+
+        statusCodes.forEach {
+            val status = it.attributes.getNamedItem("Value").textContent
+            if (status == TestCommon.SUCCESS
+                    && !TestCommon.TOP_LEVEL_STATUS_CODES.contains(status)
+                    && !TestCommon.SECOND_LEVEL_STATUS_CODES.contains(status)) {
+                throw SAMLComplianceException.createWithPropertyMessage(samlErrorCode,
+                        property = "Status Code",
+                        actual = statusCodes[0].attributes.getNamedItem("Value").textContent,
+                        expected = TestCommon.TOP_LEVEL_STATUS_CODES.toString()
+                                .replace("urn:oasis:names:tc:SAML:2.0:status:Success", "")
+                                + TestCommon.SECOND_LEVEL_STATUS_CODES.toString())
+            }
```

Comment:
❗️ This is technically not verifying the Spec exactly. I don't think we should be looking at second level status codes at all because not only are they optional but the spec says `System entities are free to define more specific status codes by defining appropriate URI references.`. i.e. those second level status codes could be anything. I'd propose these changes:
- Change line 37 to only call `node.children()` instead of `node.allChildren()`.
- Check that there's only one statusCode element that comes out of it.
- Make sure that statusCode is equal to one of the top-level error codes

Reply:
You can have multiple status codes. 
```
If the identity provider cannot or will not satisfy the request, it MUST respond 
with a <Response>message containing an appropriate error status code or 
codes.
```
I opted for doing `node.allChildren()` for that reason. I can verify that there is one `Status` element and do `node.children("Status")[0].children("Status Code)` instead.

I don't see where it says the top level status codes are mandatory and the second level codes are optional. Granted they can specify their own and we can't really check against that. I just didn't want to make the assumption that they have to specify a top-level status code.

Reply:
In Core it explains all the rules about Status Codes. Mainly section `3.2.2.2 
Element <StatusCode>`. The part that specifies that a top-level code is required:
```
Value [Required]
The status code value. This attribute contains a URI reference.
The value of the topmost <StatusCode> element MUST be
from the top-level list provided in this section.
```

As for whether or not they could have multiple top-level status codes, I'm not entirely sure based on the Core spec ☹️